File: exercise series (2019–2020)/week_10 [JavaScript]/Game of dots.py
from itertools import chain, zip_longest


class Pipopipette:
    def __init__(self, m: int, n=None):
        self.m, self.n = m, n or m
        self._score = {'A': 0, 'B': 0}
        self._player = 'A'

        self.dots = [['+'] * self.n for _ in range(self.m)]
        self.squares = [['?'] * (self.n - 1) for _ in range(self.m - 1)]
        self.hl = [['.'] * (self.n - 1) for _ in range(self.m)]
        self.vl = [['.'] * self.n for _ in range(self.m - 1)]

    def __str__(self):
        return '\n'.join(map(''.join, self.merge([self.merge(self.dots[i], self.hl[i]) for i in range(self.m)],
                                                 [self.merge(self.vl[i], self.squares[i]) for i in range(self.m - 1)])))

    @staticmethod
    def merge(a: list or tuple, b: list or tuple):
        return list(chain(*[(x, y) if y else [x] for x, y in zip_longest(a, b)]))

    def score(self):
        return tuple(self._score.values())

    def player(self):
        return self._player

    def swap_players(self):
        self._player = 'A' if self._player == 'B' else 'B'

    def claim(self, position: str):
        from re import match
        assert match("^[HV][0-9]+,[0-9]+$", position), "invalid position"
        player, scored = self.player(), False
        direction, (r, c) = position[0], map(int, position[1:].split(','))
        assert -1 < r < self.m and -1 < c < self.n, "invalid position"

        if direction == 'H':
            assert self.hl[r][c] == '.', "dots already connected"
            self.hl[r][c] = '-'
            scored = self.check(r, c, player, horizontal=True)
        elif direction == 'V':
            assert self.vl[r][c] == '.', "dots already connected"
            self.vl[r][c] = '|'
            scored = self.check(r, c, player, horizontal=False)

        if not scored:
            self.swap_players()
        return self

    def check(self, r: int, c: int, player: str, horizontal=True):
        squares_formed = 0
        for i in (-1, 1):
            if horizontal and -1 < r + i < self.m:  # check if square is formed above or below
                if (self.hl[r + i][c] == '-'
                        and self.vl[r - (i == -1)][c] == '|'
                        and self.vl[r - (i == -1)][c + 1] == '|'
                        and self.squares[r - (i == -1)][c] == "?"):
                    self.squares[r - (i == -1)][c] = player
                    squares_formed += 1
            elif not horizontal and -1 < c + i < self.n:  # check if square is formed left or right
                if (self.vl[r][c + i] == '|'
                        and self.hl[r][c - (i == -1)] == '-'
                        and self.hl[r + 1][c - (i == -1)] == '-'
                        and self.squares[r][c - (i == -1)] == "?"):
                    self.squares[r][c - (i == -1)] = player
                    squares_formed += 1

        self._score['A'] += squares_formed * (player == 'A')
        self._score['B'] += squares_formed * (player == 'B')
        return squares_formed > 0


game = Pipopipette(3)

File: exercise series (2019–2020)/week_10 [JavaScript]/Game of dots.js
class Pipopipette {
    constructor(m, n = undefined) {
        this.m = m;
        this.n = n || m;
        this.Score = [0, 0];
        this.Player = "A";

        this.dots = [...Array(this.m)].map(_ => [...Array(this.n)].map(_ => '+'));
        this.squares = [...Array(this.m - 1)].map(_ => [...Array(this.n - 1)].map(_ => '?'))
        this.hLine = [...Array(this.m)].map(_ => [...Array(this.n - 1)].map(_ => '.'))
        this.vLine = [...Array(this.m - 1)].map(_ => [...Array(this.n)].map(_ => '.'))
    }

    toString() {
        return this.merge(
            [...Array(this.m).keys()].map(i => this.merge(this.dots[i], this.hLine[i])),
            [...Array(this.m - 1).keys()].map(i => this.merge(this.vLine[i], this.squares[i]))
        ).map(row => row.join("")).join("\n");
    }

    merge(a, b) {
        return a.map((row, r) => b[r] ? [row, b[r]] : [row]).flat();
    }

    score() {
        return this.Score;
    }

    player() {
        return this.Player;
    }

    swapPlayer() {
        this.Player = this.Player === 'A' ? 'B' : 'A'
    }

    claim(position) {
        if (!position.match(/^[HV][0-9]+,[0-9]$/)) throw {name: "AssertionError", message: "invalid position"};
        let [player, scored] = [this.player(), false];
        let [direction, [r, c]] = [position[0], position.slice(1).split(",").map(i => Number(i))];
        if (r < -1 && r > this.m && c < -1 && c > this.n) throw {name: "AssertionError", message: "invalid position"};

        if (direction === "H") {
            if (this.hLine[r][c] !== ".") throw {name: "AssertionError", message: "dots already connected"};
            this.hLine[r][c] = "-";
            scored = this.check(r, c, player, true)
        } else if (direction === "V") {
            if (this.vLine[r][c] !== ".") throw {name: "AssertionError", message: "dots already connected"};
            this.vLine[r][c] = "|";
            scored = this.check(r, c, player, false)
        }
        if (!scored) this.swapPlayer();
        return this;
    }

    check(r, c, player, horizontal = true) {
        let squares_formed = 0;
        for (let i of [-1, 1]) {
            if (horizontal && r + i > -1 && r + i < this.m) {
                if (this.hLine[r + i][c] === '-'
                    && this.vLine[r - (i === -1)][c] === '|'
                    && this.vLine[r - (i === -1)][c + 1] === '|'
                    && this.squares[r - (i === -1)][c] === "?") {
                    this.squares[r - (i === -1)][c] = player;
                    squares_formed += 1;
                }
            } else if (!horizontal && c + i > -1 && c + i < this.n) {
                if (this.vLine[r][c + i] === '|'
                    && this.hLine[r][c - (i === -1)] === '-'
                    && this.hLine[r + 1][c - (i === -1)] === '-'
                    && this.squares[r][c - (i === -1)] === "?") {
                    this.squares[r][c - (i === -1)] = player;
                    squares_formed += 1;
                }
            }
        }
        this.Score[0] += squares_formed * (player === 'A')
        this.Score[1] += squares_formed * (player === 'B')
        return squares_formed > 0
    }
}